### PR TITLE
Block dungeon access

### PR DESCRIPTION
**Pull request by Gr33ndev**
Closes Gamify-IT/issues#225

### CODE REVIEW

File: Assets/Scripts/Barriers/Barrier.cs [hunk 1]
```diff
@@ -2,15 +2,25 @@
 using System.Collections.Generic;
 using UnityEngine;
 
+/// <summary>
+/// This enum defines, whether a barrier blocks access towards a dungeon or a world
+/// </summary>
+public enum BarrierType
+{
+    worldBarrier,
+    dungeonBarrier
```
Comment (maexled) [inline Assets/Scripts/Barriers/Barrier.cs:11]:
```suggestion
    WorldBarrier,
    DungeonBarrier
```

File: Assets/Scripts/GameManager/Data Classes/WorldData.cs [hunk 2]
```diff
@@ -157,5 +157,26 @@ public bool isActive()
     {
         return active;
     }
+
+    public bool dungeonIsActive(int dungeonIndex)
```
Comment (maexled) [inline Assets/Scripts/GameManager/Data Classes/WorldData.cs:161]:
```suggestion
    public bool isDungeonActive(int dungeonIndex)
```

File: Assets/Scripts/GameManager/Data Classes/DungeonData.cs [hunk 3]
```diff
@@ -103,5 +103,18 @@ public NPCData getNPCData(int index)
...
+        return active;
+    }
+
+    public void npcCompleted(int number)
+    {
+        if (number < npcs.Length)
+        {
+            npcs[number].setHasBeenTalkedTo(true);
+        }
+    }
```
Comment (maexled) [inline Assets/Scripts/GameManager/Data Classes/DungeonData.cs:118]:
```suggestion
    public void npcCompleted(int index)
    {
        if (number < npcs.Length)
        {
            npcs[index].setHasBeenTalkedTo(true);
        }
    }
```
In the overworld backend we talk about indexes

File: Assets/Scripts/GameManager/Data Classes/WorldData.cs [hunk 4]
```diff
@@ -157,5 +157,26 @@ public bool isActive()
...
+        return dungeons[dungeonIndex].isActive();
+    }
+
+    public void npcCompleted(int dungeonIndex, int number)
+    {
+        if(dungeonIndex < dungeons.Length)
+        {
+            dungeons[dungeonIndex].npcCompleted(number);
+        }
+    }
```
Comment (maexled) [inline Assets/Scripts/GameManager/Data Classes/WorldData.cs:172]:
```suggestion
    public void npcCompleted(int dungeonIndex, int index)
    {
        if(dungeonIndex < dungeons.Length)
        {
            dungeons[dungeonIndex].npcCompleted(index);
        }
    }
```

File: Assets/Scripts/GameManager/Data Classes/WorldData.cs [hunk 5]
```diff
@@ -157,5 +157,26 @@ public bool isActive()
...
+
+    public void npcCompleted(int number)
+    {
+        if(number < npcs.Length)
+        {
+            npcs[number].setHasBeenTalkedTo(true);
+        }
+    }
```
Comment (maexled) [inline Assets/Scripts/GameManager/Data Classes/WorldData.cs:180]:
```suggestion
    public void npcCompleted(int index)
    {
        if(number < npcs.Length)
        {
            npcs[index].setHasBeenTalkedTo(true);
        }
    }
```

**Review by WuestFlorian — APPROVED** [review]

File: Assets/Scripts/GameManager/GameManagerV2.cs [hunk 6]
```diff
@@ -834,6 +886,20 @@ private bool playerHasWorldUnlocked(int worldIndex)
         return false;
     }
 
+    //has player unlocked dungeon
+    private bool playerHasDungeonUnlocked(int worldIndex, int dungeonIndex)
```
Comment (maexled) [inline Assets/Scripts/GameManager/GameManagerV2.cs:890]:
```suggestion
    private bool hasPlayerDungeonUnlocked(int worldIndex, int dungeonIndex)
```

File: Assets/Scripts/GameManager/GameManagerV2.cs [hunk 7]
```diff
@@ -890,12 +956,24 @@ private void setDungeonData(int worldIndex, int dungeonIndex)
     }
     
     //mark npc as read
-    public async void markNPCasRead(string uuid)
+    public async void markNPCasRead(int worldIndex, int dungeonIndex, int number, string uuid)
```
Comment (maexled) [inline Assets/Scripts/GameManager/GameManagerV2.cs:959]:
```suggestion
    public async void markNPCAsRead(int worldIndex, int dungeonIndex, int number, string uuid)
```

**Review by cdingm — APPROVED** [review]
after talk with @WuestFlorian i understand that backend stuff is missing, for complete unlocking logic to work. The Overworld barriers work for me (they are diplayed and can be turned on and off) 
Walking into dungeons with enabled barriers is not possible

**Pull request merged**